zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2216,4 +2236,24 @@ private void validateCurrentThread() {
       throw new IllegalArgumentException("Must not be done in the zookeeper event thread.");
     }
   }
+
+  private void validateWriteSizeLimitConfig() {
+    int serializerSize = ZNRecordUtil.getSerializerWriteSizeLimit();
+    int zkClientSize = getWriteSizeLimit();
+    // ZNRecord serializer write size limit should not be set greater than size limit in ZkClient
+    if (serializerSize > zkClientSize) {
+      throw new IllegalStateException("ZNRecord serializer write size limit: " + serializerSize
+          + " is greater than size limit in ZkClient: " + zkClientSize);
+    }
+  }
+
+  private int getWriteSizeLimit() {
+    // The size margin left in default SIZE_LIMIT
+    int sizeMargin = DEFAULT_JUTE_MAXBUFFER - ZNRecord.SIZE_LIMIT;
```

Comment:
This is awkward. How about just create a constant for margin and deprecate the SIZE_LIMIT?

Reply:
ZNRecordSerializer is also using the SIZE_LIMIT to determine the actual size limit. Maybe we still need it. Otherwise, we still have `sizeLimit = JUTE_MAXBUFFER - MARGIN`. I understand this calculation is awkward, but just it is just a local variable. Or we could add one more constant for margin so we don't have to calculate it again here.

Reply:
The main concern is that you introduced some internal logic that is not configurable or commented directly where the variables are defined. Who will be able to tell the margin is derived by the default max buffer size and the default size limit?
First, this logic is not reasonable, why the margin is based on the default configs? Secondly, it introduces more chaos to our code.

Reply:
Just add a MARGIN value please.